README
Samba-LDAP version 0.03

Samba::LDAP - Manage a Samba PDC with an LDAP Backend

CPAN Port of smbldap-tools found at http://samba.idealx.org/

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


Alternatively, to install with Module::Build, use the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install



DEPENDENCIES

None.


COPYRIGHT AND LICENCE

Copyright (C) 2001-2002 IDEALX - Original smbldap-tools
Copyright (C) 2006 Suretec Systems Ltd., 
              - Gavin Henry <ghenry@suretecsystems.com>
              - http://www.suretecsystems.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
See perldoc perlgpl


Changes
Revision history for Samba-LDAP

0.03  Fri Apr 07 2006
      - Better search for group members in 
        Samba::LDAP::Group

0.02  Thu Apr 06 2006
      - Bug fix for Samba::LDAP::User

0.01  Tues Mar 28 2006
      - Bad version upload to the CPAN and might as
        well be 0.01 now ;-)

0.03_01  Mon Mar 27 2006
         - Completed preliminary documentation. 
           Samba::LDAP and Samba::LDAP::User still not complete.
         - Config files are now bundled in t/
         - Added new Machine Class

0.02_01  Fri Mar 24 2006
         - Added check for TLS connection before binding with
           masterdn and pass, as per:
           http://perlmonks.org/?node_id=538906
         - Added change_pass docs

0.01_01  Thu Mar 23 2006
         - Alpha release. Not complete. You've been warned!



scripts/test.pl
#!/usr/bin/perl -w

use strict;
use warnings;
use Carp;
use Data::Dumper;
use DateTime;
use Samba::LDAP;
use Samba::LDAP::Config;
use Samba::LDAP::User;
use Samba::LDAP::Group;
use Storable qw( nstore );

my $config = Samba::LDAP::Config->new()
   or croak "Can't create object\n";

my $smbldap = Samba::LDAP->new()
   or croak "Can't create object\n";

my $smbuser = Samba::LDAP::User->new()
   or croak "Can't create object\n";

my $smbgroup = Samba::LDAP::Group->new()
   or croak "Can't create object\n";

#print Dumper( $smbgroup );
#print Dumper( $smbldap );

#nstore $smbgroup, './test';

# Returns where smbldap.conf, smbldap_bind.conf and
# smb.conf are located
#print Dumper( $config->find_smbldap() );
#print Dumper( $config->find_smbldap_bind() );
#print Dumper( $config->find_samba() );
#print Dumper( $config );

#print "Finding SID\n";
#print Dumper( $smbldap->get_local_sid() );

#$config = $config->read_conf();
#print Dumper( $config );

#print Dumper( $smbldap->connect_ldap_master() );
#print Dumper( $smbldap->connect_ldap_slave() );

#print "Searching for Samba User: ('1' means found)\n";
#print Dumper( $smbuser->is_samba_user( 'ghenry' ) );

#print "Searching for Valid User: ('1' means found)\n";
#print Dumper( $smbuser->is_valid_user(
#'ou=Users,ou=OxObjects,dc=suretecsystems,dc=com', 'testing' ) );

#print "Getting group DN:\n";
#print Dumper( $smbgroup->_get_group_dn( 'testing' ) );

#print "Reading Group Entry:\n";
#print Dumper( $smbgroup->read_group_entry( 'testing' ) );

#print "Searching for valid Unix User: ('1' means found)\n";
#print Dumper( $smbuser->is_unix_user( 'ghenry' ) );


#my $groups_ref = [ 'staff', 'directors', 'contractors', ];
#my $groups_ref = { 
#            admin => [ 'staff', 'directors', 'contractors', ], 
#            normal => [ 'web_team', 'finance', 'cleaners', ],           
#          };
#$smbgroup->add_to_groups( $groups_ref, 'gavin' );

#print $group;

#print Dumper( $smbuser->make_hash(  clear_pass => 'testing', 
#                                    hash_encrypt_format => 'SSHA',
#                                  ) );

#print Dumper( $smbuser->_make_salt( '4' ) );

#print Dumper( $smbuser->get_next_id(
#'ou=Users,ou=OxObjects,dc=suretecsystems,dc=com', 'uidNumber' ) );

#print "Disabling User ghenry\n";
#print Dumper( $smbuser->disable_user ( 'ghenry' ) );

#print "Getting homedir for 'test'\n";
#print Dumper( $smbuser->get_homedir('test') );

#print "Disabling user 'test'\n";
#print Dumper( $smbuser->disable_user('test') ); 

#print "Deleting user 'ghenry'\n";
#print Dumper( $smbuser->delete_user( user => 'ghenry' ) ); 
#print "Waiting for 5 secs\n";
#sleep 5;


print "Checking for user 'ghenry' - 1 means they are a Samba User\n";
print Dumper( $smbuser->is_samba_user( 'ghenry' ) ); 

#my @groups = $smbgroup->find_groups( 'ghenry' );
#print "@groups\n";

#print "Adding user 'test'\n";
#print Dumper( $smbuser->add_user( 
#                                    user => 'ghenry', 
#                                    newpass => 'testing', 
#                                    windows_user => '1',
#                                    ox => '1',
#                                   
#                                ) );

#print Dumper( $smbuser->change_password ( 
#                                    user => 'ghenry',
#                                    oldpass => '{crypt}x',
#                                    newpass => 'testing',
#                                    ) );


Makefile.PL
use strict;
use warnings;

# Load the Module::Install bundled in ./inc/
use inc::Module::Install;

# Name of this distribution
name            'Samba-LDAP';

# Get most of the details from the primary module
# (The filename argument below is optional if it is placed
# consistently with the "name" above.)
all_from        'lib/Samba/LDAP.pm';

# Minimum version of Perl needed, as returned 
# by Perl::MinimumVersion 0.11
perl_version    '5.006';

# In case license is missed
license         'gpl';

# What we need to Samba::LDAP to work
requires        'Readonly'              => '1.03';
requires        'Regexp::DefaultFlags'  => '0.01';
requires        'Class::Base'           => '0.03';
requires        'Config::Tiny'          => '2.05';
requires        'Net::LDAP'             => '0.33';
requires        'Crypt::SmbHash'        => '0.12';
requires        'Unicode::MapUTF8'      => '1.11';
requires        'Digest::MD5'           => '2.36';
requires        'Digest::SHA1'          => '2.11';
requires        'MIME::Base64'          => '3.07';
requires        'List::MoreUtils'       => '0.19';

# In order to build Samba::LDAP, we need
build_requires  'Test::More'         => '0.62';
build_requires  'Test::Distribution' => '1.23';

# Install our scripts
#install_script glob('scripts/*');
    
# Auto-install all dependencies from CPAN
auto_install;

# GPG Sign Distribution
sign;

# Generate the Makefile
WriteAll;


lib/Samba/LDAP.pm
package Samba::LDAP;

# Returned by Perl::MinimumVersion 0.11
require 5.006;

use warnings;
use strict;
use Carp qw(carp croak);
use Regexp::DefaultFlags;
use Readonly;
use Net::LDAP;
use base qw( Samba::LDAP::Base );

our $VERSION = '0.03';

#
# Add Log::Log4perl to all our classes!!!!
#

# Add/Use for main code and tests:
#
# Net::LDAP::Util (DN stuff)
#
# Net::LDAP::Schema (read schemas, good to check if right ones are loaded)
#
# NET::LDAP::DSML (XML output)
#
# Net::LDAP::Extra (adding new features)

# Regexp for compiling
Readonly my $LOCALSID => qr{
                             ^SID      # Start of String
                             [ ]       # Character class for space
                             for 
                             [ ]       
                             domain
                             [ ] 
                             (\S+)     # Non-Whitespace
                             [ ]
                             is:
                             [ ] 
                             (\S+)$    # Non-Whitespace to end of line
                           };

#========================================================================
#                         -- PUBLIC METHODS --
#========================================================================

#------------------------------------------------------------------------
# get_local_sid()
#
# Tries to get the local Samba Domain SID, using:
# 'net getlocalsid'. If it fails, returns 'Can not find SID'
#------------------------------------------------------------------------

sub get_local_sid {
    my $self = shift;

    my $net_command = `LANG= \
                       PATH=/opt/IDEALX/bin:/usr/local/bin:/usr/bin:/bin \
                       net getlocalsid 2>/dev/null`;

    my ( $domain, $sid ) = $net_command =~ $LOCALSID;

    # More than likely a better way
    $self->{SID} = $sid;
    return $self->{SID} if $sid;

    # Set and return error
    $self->error('Can not find SID');
    return $self->error();
}

#------------------------------------------------------------------------
# does_sid_exist( $sid, $dn_group )
#
# Check there is a SID for user etc.
#------------------------------------------------------------------------

sub does_sid_exist {
    my $self     = shift;
    my $sid      = shift;
    my $dn_group = shift;

    my $ldap = $self->connect_ldap_master();
    my $mesg = $ldap->search(
        base   => $dn_group,
        scope  => $self->{scope},
        filter => "(sambaSID=$sid)"

#filter => "(&(objectClass=sambaSAMAccount|objectClass=sambaGroupMapping)(sambaSID=$sid))"
    );
    $mesg->code && die $mesg->error;
    return ($mesg);
}

#------------------------------------------------------------------------
# get_dn_from_line()
#
# dn = get_dn_from_line ($dn_line)
#------------------------------------------------------------------------

sub get_dn_from_line {
    my $self = shift;
    my $dn   = shift;

    # to get "a=b,c=d" from "dn: a=b,c=d"
    $dn =~ s{\A dn: [ ] }{};

    return $dn;
}

#------------------------------------------------------------------------
# do_ldapadd()
#
# Description here
#------------------------------------------------------------------------

sub do_ldapadd {
}

#------------------------------------------------------------------------
# do_ldapmodify()
#
# Description here
#------------------------------------------------------------------------

sub do_ldapmodify {
}

#------------------------------------------------------------------------
# connect_ldap_master()
#
# Connects to Master LDAP server listed in smbldap.conf. Returns Net:LDAP
# Object
#------------------------------------------------------------------------

sub connect_ldap_master {
    my $self = shift;

    # bind to a directory with dn and password
    my $ldap_master = Net::LDAP->new(
        $self->{masterLDAP},
        port    => $self->{masterPort},
        version => 3,
        timeout => 60,
      )
      or die "LDAP Error: Can't contact master ldap server ($@)";

    my $ldap_tls;
    if ( $self->{ldapTLS} == 1 ) {
        $ldap_tls = $ldap_master->start_tls(
                verify     => $self->{verify},
                clientcert => $self->{clientcert},
                clientkey  => $self->{clientkey},
                cafile     => $self->{cafile},
        );
        # Check TLS has started before binding
        $ldap_tls->code && die "Failed to start TLS: ", $ldap_tls->error;
    }
    

    my $result = $ldap_master->bind( $self->{masterDN}, password => $self->{masterPw}, );
    $result->code && die("Bind error: ", $result->error, "\n");

    return $ldap_master;
}

#------------------------------------------------------------------------
# connect_ldap_slave()
#
# Connect to Slave LDAP Directory
#------------------------------------------------------------------------

sub connect_ldap_slave {
    my $self = shift;

    my $ldap_slave = Net::LDAP->new(
        $self->{slaveLDAP},
        port    => $self->{slavePort},
        version => 3,
        timeout => 60,

      )
      or carp "LDAP error: Can't contact slave ldap server ($@)\n
	           =>trying to contact the master server\n";

    if ( !$ldap_slave ) {

        # connection to the slave failed: trying to contact the master ...
        $ldap_slave = Net::LDAP->new(
            $self->{masterLDAP},
            port    => $self->{masterPort},
            version => 3,
            timeout => 60,
          )
          or die "LDAP error: Can't contact master ldap server ($@)\n";
    }

    if ($ldap_slave) {
    my $ldap_tls;
    if ( $self->{ldapTLS} == 1 ) {
        $ldap_tls = $ldap_slave->start_tls(
                verify     => $self->{verify},
                clientcert => $self->{clientcert},
                clientkey  => $self->{clientkey},
                cafile     => $self->{cafile},
        );
        # Check TLS has started before binding
        $ldap_tls->code && die "Failed to start TLS: ", $ldap_tls->error;
    }
    

    my $result = $ldap_slave->bind( $self->{slaveDN}, password => $self->{slavePw}, );
    $result->code && die("Bind error: ", $result->error, "\n");

    return $ldap_slave;
    }

    return;
}

#------------------------------------------------------------------------
# group_type_by_name()
#
# Description here
#------------------------------------------------------------------------

sub group_type_by_name {
}

#------------------------------------------------------------------------
# list_union()
#
# Description here
#------------------------------------------------------------------------

sub list_union {
}

#------------------------------------------------------------------------
# list_minus()
#
# Description here
#------------------------------------------------------------------------

sub list_minus {
}

#========================================================================
#                         -- PRIVATE METHODS --
#========================================================================

1;    # Magic true value required at end of module
__END__

=head1 NAME

Samba::LDAP - Manage a Samba PDC with an LDAP Backend


=head1 VERSION

This document describes Samba::LDAP version 0.03


=head1 SYNOPSIS

    use Carp;
    use Samba::LDAP;
    
    my $samba = Samba::LDAP->new()
    or croak "Can't create object\n";
    my $domain = $samba->get_local_sid();


=head1 DESCRIPTION

Main functions.

B<DEVELOPER RELEASE!>

B<BE WARNED> - Not yet complete and neither are the docs!

=head1 INTERFACE 

=head2 new

Creates a new L<Samba::LDAP> object

=head2 get_local_sid

Tries to find Samba Domain SID, if not, returns undef.

=head2 does_sid_exist



=head2 get_dn_from_line



=head2 do_ldapadd



=head2 do_ldapmodify



=head2 get_user_dn2



=head2 connect_ldap_master



=head2 connect_ldap_slave



=head2 group_type_by_name



=head2 subst_configvar



=head2 read_config



=head2 read_parameter



=head2 split_arg_comma



=head2 list_union



=head2 list_minus



=head2 get_next_id



=head2 print_banner



=head2 get_domain_name



=head1 DIAGNOSTICS

None yet.


=head1 CONFIGURATION AND ENVIRONMENT

Samba::LDAP requires no configuration files or environment variables.


=head1 DEPENDENCIES

L<Carp>
L<Regexp::DefaultFlags>
L<Readonly>
L<Net::LDAP>


=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-samba-ldap@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Gavin Henry  C<< <ghenry@suretecsystems.com> >>


=head1 ACKNOWLEDGEMENTS

IDEALX for original scripts.


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2001-2002 IDEALX - Original smbldap-tools

Copyright (c) 2006, Suretec Systems Ltd. - Gavin Henry
C<< <ghenry@suretecsystems.com> >>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version. See L<perlgpl>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Samba/LDAP/Group.pm
package Samba::LDAP::Group;

# Returned by Perl::MinimumVersion 0.11
require 5.006;

use warnings;
use strict;
use Regexp::DefaultFlags;
use Readonly;
use Carp qw( croak carp );
use base qw(Samba::LDAP::Base);
use Samba::LDAP;

#use Samba::LDAP::User;
use List::MoreUtils qw( any );

our $VERSION = '0.03';

#
# Add Log::Log4perl to all our classes!!!!
#

# Our usage messages
Readonly my $ADD_TO_GROUPS_USAGE =>
  'Usage: add_to_groups( $username | HoA, Aref | HoA);';

#========================================================================
#                         -- PUBLIC METHODS --
#========================================================================

#------------------------------------------------------------------------
# is_group_member( $dn,$userid )
#
# Check that the user is a member of the group already
#------------------------------------------------------------------------

sub is_group_member {
    my $self     = shift;
    my $dn_group = shift;
    my $username = shift;

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_slave();

    my $mesg = $ldap->search(
        base   => $dn_group,
        scope  => 'base',
        filter => "(|(memberUid=$username)(member=uid=$username,$self->{usersDN}))"
    );
    $mesg->code && die $mesg->error;

    return ( $mesg->count ne 0 );
}

#------------------------------------------------------------------------
# add_to_group( $group, $username)
#
# Add $username to LDAP group $group
#------------------------------------------------------------------------

sub add_to_group {
    my $self     = shift;
    my $group    = shift;
    my $username = shift;

    my $members = q{};
    my $dn_line = $self->_get_group_dn($group);

    if ( !defined( $self->_get_group_dn($group) ) ) {
        $self->error("group $group does not exist\n");
        die $self->error();
    }

    if ( !defined($dn_line) ) {
        $self->error("Can not find group DN\n");
        die $self->error();
    }

    ( my $dn = $dn_line ) =~ s{\A dn:}{};

    # Should have been checked earlier, but check again anyway
    my $user       = Samba::LDAP::User->new();
    my $valid_user = $user->is_unix_user($username);

    # Die if they are not
    if ( $valid_user == 1 ) {
        $self->error("User $username, is not even a user on this system\n");
        die $self->error();
    }

    # Now check if the user is already present in the group
    my $is_member = $self->is_group_member( $dn, $username );
    if ( $is_member == 1 ) {
        $self->error("User $username already member of the group $group\n");
        die $self->error();
    }
    else {

        # bind to a directory with dn and password
        # It does not matter if the user already exist, Net::LDAP will add the
        # user if he does not exist, and ignore him if his already in the
        # directory.
        my $ldap = Samba::LDAP->new();
        $ldap = $ldap->connect_ldap_master();
        my $modify =
          $ldap->modify( "$dn",
            changes => [ add => [ memberUid => $username ] ] );
        $modify->code && die "failed to modify entry: ", $modify->error;
        return 0;
    }

    return 1;
}

#------------------------------------------------------------------------
# add_to_groups( $groups_ref | HoA, Aref | HoA, $username )
#
# Pass in a list of groups for the user to be added to.
#------------------------------------------------------------------------

sub add_to_groups {
    my $self     = shift;
    my $groups   = shift;
    my $username = shift;

    # Required arguments
    my @required_args = ($groups);

    # Allow HoA for adding lots of users to groups next.
    #my $groups_ref = {
    #        admin => [ 'staff', 'directors', 'contractors', ],
    #        ghenry => [ 'web_team', 'finance', 'cleaners', ],
    #      };

    croak $ADD_TO_GROUPS_USAGE
      if any { !defined $_ } @required_args;

    # Dereference the hashref passed to us by add_user in
    # Samba::LDAP::User to get the Array, or take a Aref ;-)

    my @groups;
    if ( ref($groups) eq 'HASH' ) {
        for my $key ( keys %{$groups} ) {

            # Reminder: $key is our $username
            $self->add_to_group( $key, ${$groups}{$key} );
        }
        return;
    }

    elsif ( ref($groups) eq 'ARRAY' ) {
        @groups = @{$groups};
    }
    else {
        $self->error("Need a normal Array_Ref, $ADD_TO_GROUPS_USAGE");
        croak $self->error();
    }

    for my $group (@groups) {
        $self->add_to_group( $group, $username );
    }
    return 1;
}

#------------------------------------------------------------------------
# add_group()
#
# Description here
#------------------------------------------------------------------------

sub add_group {
    my $self = shift;
}

#------------------------------------------------------------------------
# show_group( $group )
#
# Lists the entries for that group
#------------------------------------------------------------------------

sub show_group {
    my $self  = shift;
    my $group = shift;

    croak "No group specified!" if !$group;

    return $self->_read_group($group);
}

#------------------------------------------------------------------------
# list_groups()
#
# Lists the entries for that group
#------------------------------------------------------------------------

sub list_groups {
    my $self = shift;
}

#------------------------------------------------------------------------
# delete_group( $group_name )
#
# Deletes group name from LDAP tree
#------------------------------------------------------------------------

sub delete_group {
    my $self  = shift;
    my $group = shift;

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_slave();

    my $dn_line = $self->_get_group_dn($group);
    ( my $dn = $dn_line ) =~ s{\A dn: [ ] }{};

    if ( !defined($dn_line) ) {
        $self->error("$group doesn't exist\n");
        return $self->error();
    }

    my $modify = $ldap->delete($dn);
    $modify->code && croak "Failed to delete group : ", $modify->error;

    # take down session
    $ldap->unbind;

    return "$group group deleted\n.";
}

#------------------------------------------------------------------------
# read_group_entry( $group )
#
# Return all posixGroup details
#------------------------------------------------------------------------

sub read_group_entry {
    my $self  = shift;
    my $group = shift;
    my $entry;

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_slave();

    my $mesg = $ldap->search(
        base   => $self->{groupsdn},
        scope  => $self->{scope},
        filter => "(&(objectclass=posixGroup)(cn=$group))"
    );

    $mesg->code && die $mesg->error;
    my $nb = $mesg->count;

    if ( $nb > 1 ) {
        $self->error("Error: $nb groups exist \"cn=$group\"\n");

        foreach $entry ( $mesg->all_entries ) {
            my $dn = $entry->dn;
            return $dn;
        }

        return $self->error();
    }
    else {
        $entry = $mesg->shift_entry();
    }
    return $entry;
}

#------------------------------------------------------------------------
# read_group_entry_gid( $group )
#
# Read the group number in the LDAP Directory
#------------------------------------------------------------------------

sub read_group_entry_gid {
    my $self  = shift;
    my $group = shift;

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_master();

    my $mesg = $ldap->search(    # perform a search
        base   => $self->{groupsdn},
        scope  => $self->{scope},
        filter => "(&(objectclass=posixGroup)(gidNumber=$group))"
    );

    $mesg->code && die $mesg->error;
    my $entry = $mesg->shift_entry();
    return $entry;
}

#------------------------------------------------------------------------
# find_groups( $username )
#
# Find the groups that $username belongs to
#------------------------------------------------------------------------

sub find_groups {
    my $self   = shift;
    my $user   = shift;
    my @groups = ();

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_master();

    # Everything apart from Open-xchange uses memberUid, OX uses
    # member, so we do 2 searches
    my $mesg = $ldap->search(
        base   => $self->{groupsdn},
        scope  => $self->{scope},
        filter => "(&(objectclass=posixGroup)(memberUid=$user))"
    );
    $mesg->code && die $mesg->error;

    my $entry;
    while ( $entry = $mesg->shift_entry() ) {
        push( @groups, scalar( $entry->get_value('cn') ) );
    }

    # OX Search
    my $userdn = "uid=$user,$self->{usersdn}";
    my $mesg2  = $ldap->search(
        base   => $self->{suffix},
        scope  => $self->{scope},
        filter => "(&(objectclass=groupOfNames)(member=$userdn))"
    );
    $mesg2->code && die $mesg2->error;

    my $entry2;
    while ( $entry2 = $mesg2->shift_entry() ) {
        push( @groups, scalar( $entry2->get_value('cn') ) );
    }

    return (@groups);
}

#------------------------------------------------------------------------
# parse_group( $userGidNumber )
#
# Check the group is either a name or number
#------------------------------------------------------------------------

sub parse_group {
    my $self          = shift;
    my $userGidNumber = shift;

    if ( $userGidNumber =~ /[^\d]/ ) {
        my $gname  = $userGidNumber;
        my $gidnum = getgrnam($gname);
        if ( $gidnum !~ /\d+/ ) {
            return -1;
        }
        else {
            $userGidNumber = $gidnum;
        }
    }
    elsif ( !defined( getgrgid($userGidNumber) ) ) {
        return -2;
    }
    return $userGidNumber;
}

#------------------------------------------------------------------------
# remove_from_group( $group, $username )
#
# Remove the user from $group
#------------------------------------------------------------------------

sub remove_from_group {
    my $self  = shift;
    my $group = shift;
    my $user  = shift;

    my $members = q{};

    my $grp_line = $self->_get_group_dn($group);
    if ( !defined($grp_line) ) {
        return 0;
    }

    ( my $dn = $grp_line ) =~ s{\A dn: [ ] }{};

    # we test if the user exist in the group
    my $is_member = $self->is_group_member( $dn, $user );

    if ( $is_member == 1 ) {

        # delete only the user from the group
        my $ldap = Samba::LDAP->new();
        $ldap = $ldap->connect_ldap_master();
        my $modify = $ldap->modify(
            "$dn",
            changes => [
                delete => [
                    memberUid => ["$user"],
                    member    => ["uid=$user,$self->{usersdn}"],
                ],
            ]
        );
        $modify->code && die "failed to delete entry: ", $modify->error;
    }
    return 1;
}

#========================================================================
#                         -- PRIVATE METHODS --
#========================================================================

#------------------------------------------------------------------------
# _get_group_dn( $group )
#
# Searches for a groups distinguised name
#------------------------------------------------------------------------

sub _get_group_dn {
    my $self  = shift;
    my $group = shift;

    my $ldap = Samba::LDAP->new();
    $ldap = $ldap->connect_ldap_master();

    if ( $group =~ /\A \d+ \z/ ) {
        $self->{filter} =
          "(&(objectclass=posixGroup)(|(cn=$group)(gidNumber=$group)))";
    }
    else {
        $self->{filter} = "(&(objectclass=posixGroup)(cn=$group))";
    }

    my $mesg = $ldap->search(
        base   => $self->{groupsdn},
        scope  => $self->{scope},
        filter => $self->{filter},
    );
    $mesg->code && croak $mesg->error;

    for my $entry ( $mesg->all_entries ) {
        $self->{dn} = $entry->dn;
    }

    if ( !$self->{dn} ) {
        croak "Can not find $group Group";
    }

    my $dn = $self->{dn};
    chomp($dn);

    $dn = "dn: " . $dn;

    return $dn;
}

#------------------------------------------------------------------------
# _read_group( $group )
#
# Search for members of a group
#------------------------------------------------------------------------

sub _read_group {
    my $self  = shift;
    my $group = shift;

    my $ldap_slave = Samba::LDAP->new();
    $ldap_slave = $ldap_slave->connect_ldap_slave();

    my $mesg = $ldap_slave->search(
        base   => $self->{groupsdn},
        scope  => $self->{scope},
        filter => "(&(objectclass=posixGroup)(cn=$group))"
    );
    $mesg->code && croak $mesg->error;

    my $lines = '';
    for my $entry ( $mesg->all_entries ) {
        $lines .= "dn: " . $entry->dn . "\n";
        for my $attr ( $entry->attributes ) {
            {
                $lines .=
                  $attr . ": " . join( ',', $entry->get_value($attr) ) . "\n";
            }
        }
    }

    # take down session
    $ldap_slave->unbind;

    chomp $lines;
    if ( $lines eq '' ) {
        return undef;
    }

    return $lines;
}

1;    # Magic true value required at end of module

__END__

=head1 NAME

Samba::LDAP::Group - Manipulate Samba LDAP Groups


=head1 VERSION

This document describes Samba::LDAP::Group version 0.03


=head1 SYNOPSIS

    use Carp;
    use Samba::LDAP::Group;

    my $group = Samba::LDAP::Group->new()
        or croak "Can't create object\n";
    

=head1 DESCRIPTION

Various methods to add, delete, modify and show Samba
LDAP Groups

B<DEVELOPER RELEASE!>

B<BE WARNED> - Not yet complete and neither are the docs!


=head1 INTERFACE 

=head2 new

Create a new L<Samba::LDAP::Group> object

=head2 add_group

Not complete.

=head2 add_to_group

Add $username to LDAP group $group

    my $result = $group->add_to_group( $group, $username);
    print "$username added to $group\n" if $result;

=head2 add_to_groups

Pass in a list of groups for the user or users to be added to.

For one user:

    my $groups_aref = [ 'staff', 'directors', 'contractors', ];

    my $result = $group->add_to_groups( $groups_aref, $username  );
    print "$username added to groups\n" if $result;

List of users and groups:

    my $groups_ref = {
                        admin => [ 'staff', 'directors', 'contractors', ],
                        ghenry => [ 'web_team', 'finance', 'cleaners', ],
                     };

    my $result = $group->add_to_groups( $group_ref );
    print "Added to groups\n" if $result;


=head2 find_groups

Find the groups that $username belongs to. Returns an Array of groups.

    my @groups = $group->find_groups( $username );
    print "@groups";

=head2 delete_group

Deletes group name from LDAP tree

    my $delete_result = $group->delete_group( $group_name );
    print "$delete_result";

=head2 remove_from_group

Remove the user from $group. Removes C<memberUid> and C<member> entries
    
    my $result = $group->remove_from_group( $group, $username )
    print "$group removed\n" if $result;

=head2 show_group

Lists the entries for that group

    my $group_info = $group->show_group( $group );
    print "$group_info\n";

=head2 list_groups

Not complete.

=head2 parse_group

Check the group is either a name or number.

    my $result = $group->parse_group( $userGidNumber );

Not complete.

=head2 is_group_member

Check that the user is a member of the group already

    my $result = $self->is_group_member( $dn,$userid );
    print "$userid is a member of $dn\n" if $result;


=head2 read_group_entry

Return all posixGroup details. Similar to L<show_group> and will be
re-organised later

    my $group_info = $group->read_group_entry( $group );
    print "$group_info\n";

Utility method.

=head2 read_group_entry_gid

Read the group number in the LDAP Directory.

    my $group_number = $group->read_group_entry_gid( $group );
    print $group_number\n";

Utility method.

=head1 DIAGNOSTICS

None yet.


=head1 CONFIGURATION AND ENVIRONMENT

Samba::LDAP::Group requires no configuration files or environment variables.


=head1 DEPENDENCIES

L<Carp>,
L<Regexp::DefaultFlags>,
L<Readonly> and
L<List::MoreUtils>

=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-samba-ldap@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.


=head1 AUTHOR

Gavin Henry  C<< <ghenry@suretecsystems.com> >>


=head1 ACKNOWLEDGEMENTS

IDEALX for original scripts.


=head1 LICENCE AND COPYRIGHT

Copyright (c) 2001-2002 IDEALX - Original smbldap-tools

Copyright (c) 2006, Suretec Systems Ltd. - Gavin Henry
C<< <ghenry@suretecsystems.com> >>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version. See L<perlgpl>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
